### PR TITLE
transfer proposal CLI

### LINKED ISSUE #626: CLI tool to create a transfer from the guarantee fund to the distribution fund proposal
A CLI command that is for use by an authorised person that sets the amount to be transferred to the distribution (or reward) fund from the reward guarantee fund.

The output is a proposal that the Governing Board then vote on to accept/reject/abstain.

### PR DESCRIPTION
**Pull request by husio**
A command line tool for creating, signing and submitting a new proposal
for transferring funds.

I am not sure what was the expectation for the tool. After a few tries, I have modeled it similar to the [`cmd/validators](https://github.com/iov-one/weave/blob/master/cmd/validators/main.go). This tool provides as well several commands, that used in the right order allow to submit a transaction.

**The below description is now outdated.***

An example usage (tool is called `wcc` - **w**eave **c**ommand line **c**client):

```sh
Usage: ./wcc <cmd> [<flags>]

Available commands are:
        transfer-proposal
        sign
        submit
```

```sh
$ ./wcc transfer-proposal -help                                                                                                                                                       

Transfer funds from source account to the destination.

  -amount value
        An amount that is to be transferred between the source to the destination accounts. (default 1 IOV)
  -description string
        The proposal description. (default "Transfer funds to distribution account")
  -dst value
        A destination account address that the founds are send to.
  -electionrule string
        The ID of the election rule to be used.
  -memo string
        A short message attached to the transfer operation.
  -src value
        A source account address that the founds are send from.
  -title string
        The proposal title. (default "Transfer funds to distribution account")
```

Create a new transaction and store it in a file.
```sh
$ ./wcc transfer-proposal -amount "94 IOV" > /tmp/tx.binary
```

Transaction is send to whoever signature is required. The signer is using `wcc` to sign the transaction and then can send the signed transaction back or submit it.
```sh
$ ./wcc sign -key mykey123 < /tmp/tx.binary > /tmp/tx.binary.signed
$ ./wcc submit < /tmp/tx.binary.signed
```

The above commands are missing configuration flags as this is only a demo.

What do you think about this approach?

resolve #626

### CODE REVIEW

File: cmd/wcc/main.go [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+		amountFl = flCoin(fl, "amount", "1 IOV", "An amount that is to be transferred between the source to the destination accounts.")
+		memoFl   = fl.String("memo", "", "A short message attached to the transfer operation.")
+		titleFl  = fl.String("title", "Transfer funds to distribution account", "The proposal title.")
+		descFl   = fl.String("description", "Transfer funds to distribution account", "The proposal description.")
+		eRuleFl  = fl.String("electionrule", "", "The ID of the election rule to be used.")
+	)
+	fl.Parse(args)
+
+	option := app.ProposalOptions{
+		Option: &app.ProposalOptions_SendMsg{
```
Comment (alpe) [inline cmd/wcc/main.go]: What I understood from https://github.com/iov-one/weave/issues/626#issuecomment-492641972 is that there are 2 payload types that we need to support for each of the two steps.
1. [Release from escrow](https://github.com/iov-one/weave/blob/master/cmd/bnsd/app/codec.proto#L105 )
2. Cash transfer

Release from escrow is not supported

File: cmd/wcc/main.go [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+	return err
+}
+
+func flAddress(fl *flag.FlagSet, name, defaultVal, usage string) *weave.Address {
+	var a weave.Address
+	if defaultVal != "" {
+		var err error
+		a, err = weave.ParseAddress(defaultVal)
+		if err != nil {
+			panic(err.Error())
```
Comment (alpe) [inline cmd/wcc/main.go]: pp: panics won't print a nice human readable error message. Why not print and os.Exit instead?

File: cmd/wcc/main.go [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+	return &a
+}
+
+func flCoin(fl *flag.FlagSet, name, defaultVal, usage string) *coin.Coin {
+	var c coin.Coin
+	if defaultVal != "" {
+		var err error
+		c, err = coin.ParseHumanFormat(defaultVal)
+		if err != nil {
+			panic(err.Error())
```
Comment (alpe) [inline cmd/wcc/main.go]: pp: same concern about panic

**Comment by ethanfrey** [comment]
I like this approach of a multi-use command. More of the features can be added in future PRs easily.

I would make the following change requests:

* Call it `bnscli` as it is a bns-specific client
* Add it to the root level Makefile (at least `make install`), and some docs... similar request as for cleanproto - I understand not doing it yet before feedback, but good to do before merge.
* 3 separate steps is great for multi-sig support, but it would be nice to allow them to run as one step as well. I will explain a proposal below, but some UX enhancement there would be good. (This may also be a future PR if you want to merge in small steps).
* (in a future PR) merge the `cmd/validators` functionality in here and remove `cmd/validators`

**Comment by ethanfrey** [comment]
1. sign and submit are linked by the last signer on multi-sig. Maybe allow this:

`./wcc sign -key mykey123 < /tmp/tx.binary > /tmp/tx.binary.signed`
OR
`./wcc sign -key mykey123 -submit < /tmp/tx.binary`

(and require all submit-specific flags, like node endpoint, if the -submit flag is present)

2. When creating a proposal, signing is almost always a required step, so we could add:

`./wcc transfer-proposal -amount "94 IOV" -key mykey123 > /tmp/tx.binary.signed`
to the existing:
`./wcc transfer-proposal -amount "94 IOV" > /tmp/tx.binary`

(presence of `-key` means sign along with creation).

3. If this is not multi-sig, we will want to do all three steps, so creation should support `-submit` flag, just like signed (1 above). This is valid only if `-key` is present:

`./wcc transfer-proposal -amount "94 IOV" -key mykey123 -submit` will be a one-shot stop to do it all

In summary, we have these paths:

`create` -> unsignedTx
`sign` + unsignedTx -> signedTx
`submit` + signedTx -> blockchainResult

I want to add:

`create-and-sign` -> signedTx
`sign-and-submit` + unsignedTx -> blockchainResult
`create-sign-and-submit` -> blockchainResult

but not adding more commands, just usage of optional flags

File: cmd/wcc/main.go (hunk 1 above)
Comment (husio) [reply cmd/wcc/main.go]: Good catch! :+1:

File: cmd/wcc/main.go (hunk 2 above)
Comment (husio) [reply cmd/wcc/main.go]: I do not have a strong opinion about it. My thinking was that this (panic + err) will provide both an error message and the stack trace of where it happened. I am using `panic` only in truly unusual situations.

I can change it to `fmt.Fprintf(os.Stderr, "%+v", err)` so that if supported, the error stack trace will be printed instead. I think for the user it does not make the difference. It matters only for us programmers.
What do you think?

**Comment by husio** [comment]
@ethanfrey I like your ideas and I think this is the right direction. I think there is much more  that can be improved. For example, all the time that `-key` should not be optional and a file `$HOME/.bnsd.key` or something similar should be checked instead (passing secrets via the command line argument is wrong :see_no_evil:).
Another improvement is to merge this with `cmd/validators`.

I think writing this tool might be a longer process, going beyond resolving issue #626. I will focus for now on basics and providing the requested functionality. Once I get this right I will work on making the UI good. After that it will be time to merge with `cmd/validators`.

**Comment by ethanfrey** [comment]
> I think writing this tool might be a longer process, going beyond resolving issue #626. I will focus for now on basics and providing the requested functionality. Once I get this right I will work on making the UI good. After that it will be time to merge with cmd/validators.

Yes, I agree. Let's get a nice UI for this tool (which may be a few PRs as mentioned above... one for getting it working, another for getting a nice UI, another for better key management), then some for merging with validators cli.. etc.

**Comment by ethanfrey** [comment]
> @ethanfrey do you think that the is an acceptable improvement to the flow?
The is no change in the tool itself. Instead of running steps separately I am combining them with a pipe.

`bsncli transfer-proposal -amount "94 IOV" | bnscli sign -key mykey123 | bnscli submit`

Very nice, I didn't realize this worked, but of course - unix pipes ftw!

I like this idea, simple and powerful. Just make sure to document this as a typical workflow to combine steps.

**Comment by ethanfrey** [comment]
@husio I would suggest addressing code changes as requested by @alpe and finishing up some of the unimplemented flags how it works, then the following minor changes in this PR:

* Call it bnscli as it is a bns-specific client (rename folder)
* Add it to the root level Makefile (at least make install)... similar request as for cleanproto - I understand not doing it yet before feedback, but good to do before merge.
* And some docs on how to use it (compile then run as normal use case, also give examples of using pipes for some common workflows - single-tx, preping multisig, submitting multisig with last signature)

That should be plenty for this issue, can add more in other PRs.

File: cmd/wcc/main.go (hunk 2 above)
Comment (ethanfrey) [reply cmd/wcc/main.go]: Since this is just used in main.go, how about adding a `func die(err error)` which will end the program. And call this everywhere we abort.

Then we can discuss this in one place. Either

```go
func die(err error) {
  panic(err)
}
```

```go
func die(err error) {
  panic(err.Error())
}
```

```go
func die(err error) {
  fmt.Fprintf(os.Stderr, "%+v", err)
  os.Exit(1)
}
```

I kind of like the last one, but I think just having one function for termination will make this easy to change later if needed.

**Comment by ruseinov** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry. <br />
You can find it at <a href="https://github.com/iov-one/weave/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by husio** [comment]
I have updated the code.

- command is renamed to `cmd/bnscli`
- flag parsing panic was changed to use `os.Exit(2)` as it makes more sense
- [a new command](https://github.com/iov-one/weave/pull/695/commits/0ecf60cb0ab3c677694adb0609ff72a3334330e6#diff-7add438e24ba4fb2be23de3fc9e20471R22) for escrow releasing proposal as added (missing functionality)
- the code was split into more files by the functionality it implements
- small comments improvements

Please have a look at let me know if all is going in the right direction. The change is hard to track because of the splitting into files. I have not changed much though.

In the meantime, I will write tests and make sure that only the basics are implemented. All further improvements are planned for other pull requests.

File: cmd/bnscli/cmd_gov.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
+package main
```
Comment (ethanfrey) [inline cmd/bnscli/cmd_gov.go]: When building other cli tools with similar shape (eg. bnsd, gaiacli), I used a `commands` subdirectory with all the commands (and flag parsing). This did allow some (minimal) testing of that code.

The `main` package would just be one file that would import these commands and grab the first argument off the cli and dispatch down to the proper command.

I found cobra/viper to provide quite a few helpers, but also be heavy, opinionated libraries. If you wish to pull in some libraries beyond stdlib to make cli dev easier, I'm very open to that

Comment (ethanfrey) [reply cmd/bnscli/cmd_gov.go]: If they are really generic, they could go into `commands/client`, next to `commands/server`.
If they are bnscli specific, then `cmd/bnscli/commands` is a good place

File: cmd/bnscli/cmd_sigs.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+Sign given transaction. This is decoding a transaction data from standard
+input, adds a signature and writes back to standard output signed transaction
+content.
+
+`)
+		fl.PrintDefaults()
+	}
+	var (
+		tmAddrFl = fl.String("tm", "https://bns.NETWORK.iov.one:443", "Tendermint node address. Use proper NETWORK name.")
+		keyFl    = fl.String("key", "", "Hex encoded, private key that transaction should be signed with.")
```
Comment (ethanfrey) [inline cmd/bnscli/cmd_sigs.go]: At the least, we should refer to a file here.

https://github.com/iov-one/weave/blob/master/cmd/bnsd/client/keys.go

has some basic functions for working with private keys on disk. Not encrypted or secure yet, but a good place to get this working. We could add an encryption function and password here in the future (I did this for gaia/cosmos-sdk, and worked on the iov-core model, so I can give a number of pointers). Out of scope for this PR, but simply moving to unencrypted file storage (which can be mode 0400 and not visible via `ps`) will be a big security improvement

File: cmd/bnscli/cmd_sigs.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+		fmt.Fprint(flag.CommandLine.Output(), `
+Sign given transaction. This is decoding a transaction data from standard
+input, adds a signature and writes back to standard output signed transaction
+content.
+
+`)
+		fl.PrintDefaults()
+	}
+	var (
+		tmAddrFl = fl.String("tm", "https://bns.NETWORK.iov.one:443", "Tendermint node address. Use proper NETWORK name.")
```
Comment (ethanfrey) [inline cmd/bnscli/cmd_sigs.go]: It is kind of nice to allow defining `-tm` via environmental variable (or config file), as this would be used by all invocations. The script wrappers you proposed could have something like `export TM_ADDR=${TM_ADDR:-https://...}.` on the top and then not worry about the flags and be very easy to customize.

Comment (ethanfrey) [reply cmd/bnscli/cmd_sigs.go]: This environmental variable technique is used here:
https://github.com/iov-one/weave/blob/master/cmd/bnsd/main.go#L22-L23

File: cmd/bnscli/cmd_submit.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	fl.Usage = func() {
+		fmt.Fprint(flag.CommandLine.Output(), `
+Read binary serialized transaction from standard input and submit it.
+
+Make sure to collect enough signatures before submitting the transaction.
+`)
+		fl.PrintDefaults()
+	}
+	var (
+		tmAddrFl = fl.String("tm", "https://bns.NETWORK.iov.one:443", "Tendermint node address. Use proper NETWORK name.")
```
Comment (ethanfrey) [inline cmd/bnscli/cmd_submit.go]: I like this, same comments as above (subdir - probably generic, and TM_ADDR environmental variable)

File: cmd/bnscli/main.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+// this program. The name is used to match with the first argument given.
+//
+// When a cmd function is callend it is given stdin, stdout and command line
+// arguments except the program name and this command name. It is the
+// responsibility of the command function to parse the arguments. Use os.Stderr
+// to write error messages.
+//
+// TODO - document here how to create a command function. They all must follow
+// the same convention. Ie use flag package and provide help information.
+var commands = map[string]func(input io.Reader, output io.Writer, args []string) error{
```
Comment (ethanfrey) [inline cmd/bnscli/main.go]: Nice way to avoid long verbose switch statement :+1:

File: cmd/bnscli/cmd_sigs.go (hunk 6 above)
Comment (husio) [reply cmd/bnscli/cmd_sigs.go]: Done

File: cmd/bnscli/cmd_submit.go (hunk 7 above)
Comment (husio) [reply cmd/bnscli/cmd_submit.go]: Done

File: cmd/bnscli/cmd_sigs.go (hunk 5 above)
Comment (husio) [reply cmd/bnscli/cmd_sigs.go]: Done.

File: cmd/bnscli/cmd_gov.go (hunk 4 above)
Comment (husio) [reply cmd/bnscli/cmd_gov.go]: Can you explain a bit more what is the advantage of using a subdirectory? The current implementation is all in `main` package and I am able to test it.
None of the functionality is generic. They rely heavily on using a command line interface and they are complete in their functionality. I do not think we will need this code anywhere else.

I was wondering about using a command line library. As always I am very cautious of adding a dependency. I hope I can implement all I need without importing big dependencies. Only if the advantage is huge I will switch :smirk:

**Comment by husio** [comment]
I have addressed all comments but the subdirectory one. I have also added tests for all functionality except the transaction creation (to be done). Once all unit tests are there, I will write an "acceptance test", that will test the tool using command line :crossed_fingers:

File: cmd/wcc/main.go (hunk 2 above)
Comment (husio) [reply cmd/wcc/main.go]: I have implemented [`flagDie` function](https://github.com/iov-one/weave/blob/cli_create_transfer_issue_625/cmd/bnscli/flag.go#L85). It does `printf` and `exit 2`. Formatting of the message is left to the caller, but you cannot do much more than just print the error. Error formatting doesn't matter as they all come most of the time directly from the stdlib.
